### ISSUE TITLE
Persist multipart uploads in backend

### ISSUE THREAD

**Issue by T4cC0re** [johannesboyne/gofakes3#33]
Hey,

I think it would be a great idea to persist multipart uploads into the backend. (I kinda need that for a project I am working on) Currently implementing the following: (But not PR worthy yet)

Make the backends store chunks of the content, rather than the whole file. (As an optional interface maybe?)  
This way we can just keep adding chunks to the backend object until we are done (maybe have a ready/finished flag on the object, too. Not sure, yet)

Then when a download is requested we can use a io.MultiReader to read out the chunks as we see fit. (maybe some clever seeking technique to not waste resources/file-handles)

**Comment by shabbyrobe** [comment]
This would be great!! In-memory multipart uploads were just a shortcut I took for my own employer's use case, I've been meaning to circle back to them for a while. My apologies if you were bitten by this limitation.

> Make the backends store chunks of the content, rather than the whole file. (As an optional interface maybe?)

An optional interface sounds ideal. I would like the existing "in memory if your backend doesn't support it" feature to remain so users of backends that don't implement it still get access to a version of the feature (possibly behind an opt-in flag). We've been slowly moving to more granular optional interfaces and I think it's going well so far. There's a PR in flight here that may influence your PR a litte bit (#29) - I will clean up the remaining bits to get it merged as soon as I can.

> Make the backends store chunks of the content

Just for the case of storing the parts of the multipart upload? Or as a modification to the method that handles putting an object? Which backend are you using?

Johannes is the final arbiter for this stuff, but so far we've been taking a "near-enough-is-good-enough" approach to the accuracy of the S3 support when merging new features, then refining as we go. If what you have advances support but isn't perfect and doesn't limit use-cases already supported (and you don't mind things moving around a little bit later if we need to restructure things a bit to make room for the next set of features), I'd vote for us to accept it with rough edges. The listing/pagination stuff in particular is quite fussy; I punted on it when preparing #30 with the intention of circling back to making it accurate as a stand-alone piece of work. Some of the backends don't currently support listing or pagination at all and just return the whole set. It's fine to mimic this if necessary.

Alternatively, if you get something part-way done, run out of time, move on to other work (or other, more featureful S3 implementations), etc, I'd be grateful if you would fire up what you have as a WIP PR anyway and we can pick it up and run with it.

Thanks for reaching out!

**Comment by shabbyrobe** [comment]
Ah I see you already have a branch in flight that covers most of what I said! Sorry, I should've done some more digging first. 😳 

It looks like you're storing the multipart uploads as regular objects. S3 provides separate endpoints for listing multiparts and regular objects - I'm not sure we should conflate the two in the Backend interface.

**Comment by T4cC0re** [comment]
It might be a good idea to state the background of what I want to accomplish.

I found gofakes3 as I wanted to implement some kind of S3 adapter, and it seems very interesting because of the pluggable backends and the general embedabilty of gofakes3.  
My intention is to do as little read & write operations as necessary during this.

So my initial though was to just make every object 'chunked'. This has the benefit, that we can read multiple readers with  a io.MultiReader (as mentioned above + I already implemented a closing variant) and regular uploads, that are not multipart can simply just provide a single reader (=chunk).

That way we do have a break in the API, but it makes a GET and HEAD onto objects simpler.

The only thing I am not quite sure about how to handle is the in-flight state of that object.  
For that purpose I added the `IsFinished` property onto the `Object` to make sure we can ignore those while listing Object until the upload of that key has been marked as finished by the uploader.

I can see 2 problems with this though:

1. There could be multiple upload IDs geared towards the same key, which wins?
2. A backend might store the obkect key in a map-like structure, so there con only be a single `Object` for a key. That would be unfinished during the upload and thus not listed.

After comtemplating the above, I still think the slice of `ReadCloser`s is a good idea to keep it simple, but I think I need to extend the `MultipartBackend` to have some kind of 'staging area' where we can dump the uploads to, and then later transfer that into the 'real' state.  
One thing I really want to avoid is to read all the data from the backend to save it via the `PutObject(bucketName, key string, meta map[string]string, input io.Reader, size int64)` method.

I'll try to come up with something here.

**Comment by johannesboyne** [comment]
Hi @T4cC0re contributions are always welcome thanks a lot for raising the topic, which is a quite interesting one! I'm very interested to see a first draft implementation.
Blake (@shabbyrobe) with all his contributions to this repo, has done an incredible good job with the pluggable backends, nice to read it was the reason you chose `gofakes3` to build upon.

**Comment by T4cC0re** [comment]
I went ahead an implemented the first usage of the new Interface. No backend code yet, but I found some stuff I have not been aware of before by doing that.

Next step for me is to enhance the `bolt` backend to use it.

I rolled back my change about the `Object.Contents` being a `[]io.ReadCloser`. Instead I think this should get handled by the backend transparently by exposing a MultiReadCloser if desired.

I shifted from gofakes3 doing the most work, to moving the complexity to the backend, because ultimately this is not easy to abstract in a way that is usable for a large number of use cases.

I'll add a WIP PR right away. I think this is mostly what I wanna go with (with maybe some small adjustments)
